pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -131,10 +144,22 @@ public void addSegment(ImmutableSegment segment) {
       return;
     }
 
-    Preconditions.checkArgument(segment instanceof ImmutableSegmentImpl,
-        "Got unsupported segment implementation: {} for segment: {}, table: {}", segment.getClass(), segmentName,
-        _tableNameWithType);
-    addSegment((ImmutableSegmentImpl) segment, new ThreadSafeMutableRoaringBitmap(), getRecordInfoIterator(segment));
+    Lock segmentLock = SegmentLocks.getSegmentLock(_tableNameWithType, segmentName);
+    segmentLock.lock();
```

Comment:
Will this work? Don't we already take a lock on same segment in `HelixInstanceDataManager.addOrReplaceSegment` in which this method gets called?

Reply:
Yes, the same lock can be acquired by the same thread twice (reentrant lock). Adding this lock can prevent add and remove happening at the same time

Reply:
cool.
Yeah the rest of it seems fine.